### PR TITLE
Improvement: Additional Raft Testing

### PR DESCRIPTION
**Pull request by faineance**
- create RaftTestT from RaftTestM, to be able to use RaftTest without ConcIO ( dejafu )
- basic tests simulating follower catchup in cases of conflicts and nodes being behind
- fix deadlock with leader never responding to client reads when the client read was included in an append entry response with aerSuccess as false

### CODE REVIEW

File: src/Raft.hs [hunk 1]
```diff
@@ -382,19 +381,21 @@ handleAction
   -> RaftT v m ()
 handleAction action = do
   logDebug $ "Handling [Action]: " <> show action
+  let sendRPCThread nid rpcMsg =
```
Comment (tdietert) [inline src/Raft.hs]: Can we move this down to the where clause of `handleAction`? Nice job abstracting out the repetition.

File: src/Raft/Leader.hs [hunk 2]
```diff
@@ -46,31 +46,45 @@ handleAppendEntries (NodeLeaderState ls)_ _  =
   pure (leaderResultState Noop ls)
 
 handleAppendEntriesResponse :: forall sm v. RPCHandler 'Leader sm AppendEntriesResponse v
-handleAppendEntriesResponse ns@(NodeLeaderState ls) sender appendEntriesResp
-  -- If AppendEntries fails (aerSuccess == False) because of log inconsistency,
-  -- decrement nextIndex and retry
-  | not (aerSuccess appendEntriesResp) = do
+handleAppendEntriesResponse ns@(NodeLeaderState ls) sender appendEntriesResp = do
+  newLeaderState <- if aerSuccess appendEntriesResp
```
Comment (tdietert) [inline src/Raft/Leader.hs]: This is a bit weirdly formatted. Can we do:
```
if <predicate>
  then <then branch>
  else <else branch>
```

File: src/Raft/Leader.hs [hunk 3]
```diff
@@ -46,31 +46,45 @@ handleAppendEntries (NodeLeaderState ls)_ _  =
...
-              newNextIndices = Map.insert sender (lastLogEntryIdx + 1) (lsNextIndex ls)
-              newMatchIndices = Map.insert sender lastLogEntryIdx (lsMatchIndex ls)
-              lsUpdatedIndices = ls { lsNextIndex = newNextIndices, lsMatchIndex = newMatchIndices }
-          -- Increment leader commit index if now a majority of followers have
-          -- replicated an entry at a given term.
-          lsUpdatedCommitIdx <- incrCommitIndex lsUpdatedIndices
-          when (lsCommitIndex lsUpdatedCommitIdx > lsCommitIndex lsUpdatedIndices) $
-            updateClientReqCacheFromIdx (lsCommitIndex lsUpdatedIndices)
-          pure lsUpdatedCommitIdx
-        Just n -> handleReadReq n ls
```
Comment (tdietert) [inline src/Raft/Leader.hs]: I'm not sure if this is the right thing to do here. 
If the `aerReadRequest` field is not `Nothing`, that means that the Append Entries RPC sent by the leader, to which this Append Entries Response is a response, was sent in response to a client read request, and contained no entries. Thus, the above logic (lines 60 - 71) are unnecessary (see: https://github.com/adjoint-io/raft/blob/master/src/Raft/Leader.hs#L123).

File: test/TestNodeState.hs [hunk 4]
```diff
@@ -0,0 +1,5 @@
+module TestNodeState where
```
[the remark this answers is not in the document]
Comment (tdietert) [reply test/TestNodeState.hs]: Yeah this file should probably be removed.

File: test/TestDejaFu.hs [hunk 5]
```diff
@@ -434,68 +135,80 @@ comprehensive eventChans clientRespChans =
...
 
-heartbeat :: TestEventChan -> ConcIO ()
-heartbeat eventChan = do
-  sysTime <- liftIO getSystemTime
-  atomically $ writeTChan eventChan (TimeoutEvent sysTime HeartbeatTimeout)
+test_AEFollowerBehindOneTerm :: TestTree
+test_AEFollowerBehindOneTerm =
+  testGroup "AEFollowerBehindOneTerm"
+    [ majorityNodeStatesEqual (pure ())
+      [ (node0, Term 4, SampleEntries.entries)
```
[the remark this answers is not in the document]
Comment (tdietert) [reply test/TestDejaFu.hs]: @Acentelles that would indeed be a good idea, but I think it would take a long time to implement. We should add it to the TODO list.

File: test/TestDejaFu.hs [hunk 6]
```diff
@@ -434,68 +135,80 @@ comprehensive eventChans clientRespChans =
...
+  let (inAgreement, inDisagreement) = partition
+        (== testNodeLog (testStates Map.! node0))
+        (fmap testNodeLog $ Map.elems testStates)
+  in  (length inAgreement) > length inDisagreement
+correctResult (Left _) = False
+
+test_AEFollower :: TestTree
+test_AEFollower =
+  testGroup "AEFollower"
+    [ majorityNodeStatesEqual (syncClientWrite node0 (Set "x" 7))
```
[the remark this answers is not in the document]
Comment (tdietert) [reply test/TestDejaFu.hs]: @Acentelles in this PR or in another? Do you think it's good for now?

File: test/TestDejaFu.hs [hunk 7]
```diff
@@ -434,68 +135,80 @@ comprehensive eventChans clientRespChans =
...
-    _ -> panic "Failed to receive client write response..."
+-- | Check if the majority of the node states are equal after running the given RaftTestClientM
+-- program
+-- TODO hardcoded to running with 3 nodes
+majorityNodeStatesEqual :: RaftTestClientM a -> TestNodeStatesConfig -> TestTree
+majorityNodeStatesEqual clientTest startingStatesConfig  =
+  testDejafusWithSettings dejaFuSettings
+    [ ("No deadlocks", deadlocksNever)
+    , ("No exceptions", exceptionsNever)
+    --, ("always same end result", alwaysSame)
```
[the remark this answers is not in the document]
Comment (tdietert) [reply test/TestDejaFu.hs]: @faineance We should either explain why these are commented out, or removed the commented out lines.

File: test/TestDejaFu.hs [hunk 8]
```diff
@@ -434,68 +135,80 @@ comprehensive eventChans clientRespChans =
...
+      , (node1, Term 2, Seq.take 6 SampleEntries.entries)
+      , (node2, Term 2, Seq.take 6 SampleEntries.entries)
+      ]
+    ]
 
-clientReadRespChan :: RaftTestClientM (ClientResponse Store StoreCmd)
-clientReadRespChan = do
-  clientRespChan <- lift (asks testClientEnvRespChan)
-  lift $ lift $ atomically $ readTChan clientRespChan
+--prop_AEFollowerBehind :: Small (Positive Integer) -> Small (Positive Integer) -> Bool
```
Comment (tdietert) [inline test/TestDejaFu.hs]: Why is this function commented out?

File: test/TestDejaFu.hs [hunk 9]
```diff
@@ -13,293 +13,34 @@
...
-    nid = raftConfigNodeId (testRaftNodeConfig testEnv)
-    Just eventChan = Map.lookup nid (testNodeEventChans testEnv)
-    dummyTimer = pure ()
-
-forkTestNodes :: [TestNodeEnv] -> TestNodeStates -> ConcIO [ThreadId ConcIO]
-forkTestNodes testEnvs testStates =
-  mapM (fork . flip runTestNode testStates) testEnvs
-
---------------------------------------------------------------------------------
+dejaFuSettings = defaultSettings { _way = randomly (mkStdGen 42) 100 }
```
Comment (tdietert) [inline test/TestDejaFu.hs]: Why was this removed from the `where` clause of the `testConcurrentProps` function?

File: test/RaftTestT.hs [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+askSelfNodeId = asks (raftConfigNodeId . testRaftNodeConfig)
+
+lookupNodeEventChan :: forall m. MonadConc m => NodeId -> RaftTestT m (TestEventChan m)
+lookupNodeEventChan nid = do
+  testChanMap :: TestEventChans m <- asks testNodeEventChans
+  case Map.lookup nid testChanMap of
+    Nothing -> throwTestErr $ "Node id " <> show nid <> " does not exist in TestEnv"
+    Just testChan -> pure testChan
+
+getNodeState :: MonadConc m => NodeId -> (RaftTestT m) TestNodeState
```
Comment (tdietert) [inline test/RaftTestT.hs]: the `(RaftTestT m)` does not need parens around it.

File: test/TestDejaFu.hs (hunk 9 above)
Comment (faineance) [reply test/TestDejaFu.hs]: because it's also used for the other deja fu tests

File: test/TestDejaFu.hs (hunk 8 above)
Comment (faineance) [reply test/TestDejaFu.hs]: was a sketch of how using quickcheck we might be able to test with random amounts of entries, and nodes behind random amounts

File: test/TestDejaFu.hs (hunk 9 above)
Comment (tdietert) [reply test/TestDejaFu.hs]: Hmm, maybe we should give each dejafu test "ownership" over their own `dejafuSettings` value, such that later on when we are fiddling with the memory models, scheduling types, etc, we won't be globally changing the test runners, only locally. I suppose this is fine for now though.

File: test/TestDejaFu.hs (hunk 8 above)
Comment (faineance) [reply test/TestDejaFu.hs]: a wip attempt that I never finished

Comment (tdietert) [reply test/TestDejaFu.hs]: Generally I'm ok with leaving things commented out as long as there is an explanation above the code describing _why_ it is commented out.
